hardware_interface/src/resource_manager.cpp

```diff
@@ -1283,6 +1307,56 @@ HardwareReadWriteStatus ResourceManager::read(
   return read_write_status;
 }
 
+void ResourceManager::publish_all_interface_values() const
+{
+  control_msgs::msg::DynamicInterfaceValues interface_values;
+  interface_values.header.stamp = resource_storage_->clock_interface_->get_clock()->now();
+
+  control_msgs::msg::InterfaceValue state_interface_values;
+  for (const auto & state_interface_name : resource_storage_->available_state_interfaces_)
+  {
+    if (const auto iter = resource_storage_->state_interface_map_.find(state_interface_name);
+        iter != resource_storage_->state_interface_map_.end())
+    {
+      state_interface_values.interface_names.push_back(state_interface_name);
+      state_interface_values.values.push_back(iter->second.get_value());
+    }
+    else
+    {
+      RCUTILS_LOG_WARN_NAMED(
+        "resource_manager",
+        "State interface '%s' is in available in state_interface_map_, but could not get the "
+        "interface",
+        state_interface_name.c_str());
+    }
+  }
+
+  control_msgs::msg::InterfaceValue command_interface_values;
+  for (const auto & command_interface_name : resource_storage_->available_command_interfaces_)
+  {
+    if (const auto iter = resource_storage_->command_interface_map_.find(command_interface_name);
+        iter != resource_storage_->command_interface_map_.end())
+    {
+      command_interface_values.interface_names.push_back(command_interface_name);
+      command_interface_values.values.push_back(iter->second.get_value());
+    }
+    else
+    {
+      RCUTILS_LOG_WARN_NAMED(
+        "resource_manager",
+        "command interface '%s' is available in command_interface_map_, but could not get the "
+        "interface",
+        command_interface_name.c_str());
+    }
+  }
+  interface_values.states = state_interface_values;
+  interface_values.commands = command_interface_values;
+
+  rt_interface_values_publisher_->lock();
```

Comment:
I think this can block the RT loop quite a lot.
https://github.com/ros-controls/realtime_tools/blob/0ce9f77d1cf693a5340a2a986a461a09a0f04514/include/realtime_tools/realtime_publisher.h#L133-L143

We could use `trylock()`, still we wouldn't be able to refresh data every time. But it is better approach.